rust/src/dataset.rs

```diff
@@ -487,6 +487,27 @@ impl Dataset {
         Ok(counts.iter().sum())
     }
 
+    /// Count the number of rows in the dataset with filter.
+    pub async fn count_rows_with_filter(&self, expression: &str) -> Result<usize> {
+        // Open file to read metadata.
+        let count = stream::iter(self.get_fragments())
+            .map(|f| async move {
+                f.scan()
+                    .filter(expression)
+                    .map_err(Error::from)?
+                    .try_into_stream()
+                    .await
+                    .map_err(Error::from)?
+                    .try_fold(0usize, |acc, _| async move { Ok(acc + 1) })
```

Comment:
This is not clear to me how does this work. Is it doing a counting of each "returned stuff from the stream"?

A few concerns here:

* We might want to limit the read only to the columns that satisfy the filter, to make things faster.
* Fragment should have a `Fragment::count_rows(expr: Optional<&str>)` to handle internal details?
* The stream returned from `f.scan.try_into_stream()` is `Stream<item=Result<RecordBatch>>`, IIUC, here the code is counting the number of batches. instead of the number of rows.

Reply:
About two last concerns I got your meaning . but can you tell more about the first one ?

Reply:
Sure thing, as Lance is a columnar format, we'd like to limit the I/O as little as possible. Many of our use cases have very big columns, i.e., storing image, sensor data and etc, but usually `count_rows(expr)` expression can be applied to small columns, i.e., 2-3 numeric or string columns. So what we can do here to make sure we don't waste any I/O is that, we only read the columns that are referred in the filter expression.

Reply:
and how do you think it should be consider to do ? is there any similar issue in repo or any reference ?

Reply:
Yes

https://github.com/lancedb/lance/blob/a141d66ddeb26543b06531b088a00a8644873ee7/rust/src/dataset/scanner.rs#L408 

Here is an example

Reply:
you mean something like this ?